### ISSUE TITLE
FocusGained event should clear each buffers' 'changedtick' variable in the cache

### ISSUE THREAD

**Issue by johnmiked15** [airblade/vim-gitgutter#533]
First, just wanted to thank you for a great plugin. I rely on it everyday and it's awesome!

There is one scenario that I think could be improved, however. When `FocusGained` triggers, it ends up calling `gitgutter#all(1)`:

https://github.com/airblade/vim-gitgutter/blob/a986ab054788776dca269d6c289b470255d54e8c/autoload/gitgutter.vim#L5-L13

Unfortunately, `tabpagebuflist()` only returns the buffers that are currently associated in visible windows on the current tab. It doesn't return *every* buffer. This is necessary when you have many files that have changes, and you git commit them in a different window. When coming back to vim and switching to a buffer that wasn't visible (that had signs last time it was open), the signs are not updated.

Instead of actually doing the diff for every buffer, is it possible to augment the above function to also clear the 'changedtick' bufvar for hidden buffers when called? This would allow `gitgutter#process_buffer()` to update the signs the next time the buffer is loaded after a `FocusGained` event (since `s:has_fresh_changes()` would return true).

**Comment by airblade** [comment]
Glad it helps!

> When coming back to vim and switching to a buffer that wasn't visible (that had signs last time it was open), the signs are not updated.

They should be.  When `FocusGained` triggers, after updating the current tab's buffers via `gitgutter#all(1)`, it [sets a flag](https://github.com/airblade/vim-gitgutter/blob/a986ab054788776dca269d6c289b470255d54e8c/plugin/gitgutter.vim#L219) called `gitgutter_force` on every other tab.

Then, when you enter one of those other tabs, the [flag makes gitgutter force-update](https://github.com/airblade/vim-gitgutter/blob/a986ab054788776dca269d6c289b470255d54e8c/plugin/gitgutter.vim#L190-L195) the tab's buffers.

What commit is your plugin installation at?  Also, can you confirm the `FocusGained` event is in fact firing?

**Comment by johnmiked15** [comment]
Plugin is at current HEAD. `FocusGained` events are confirmed to be firing appropriately.

The problem I believe is with how `gitgutter#all()` works. The flag you’re referring to will call that function when a hidden *tab* is activated, but `tabpagebuflist()` still only iterates through visible buffers. So when activating a hidden buffer, the only thing that’s called is `gitgutter#process_buffer()` with force set to false, because the flag has been cleared and it didn’t iterate through every buffer when it was force updating. 

My suggestion of clearing the ‘changedtick’ buffer var cache would just allow `s:has_fresh_changes()` to return true and allow the non-forced `gitgutter#process_buffer()` to rerun the diff when hidden buffers are activated after a `FocusGained` event fires.

**Comment by johnmiked15** [comment]
I can work on a set of reproducible steps for you, but in the mean time, I did notice you already have a helper function `s:clear(bufnr)` that I think just needs to be called for *every* buffer (not just the ones returned by `tabpagebuflist()`). I also think this same problem occurs when disabling gitgutter, since that same function is used to determine the buffers from which to clear signs (which only gets visible buffers).

**Comment by airblade** [comment]
You're right, there was a gap in my thinking.  I was tacitly assuming that every buffer would be visible in a tab – which of course is nonsense.  I don't know why I made that mistake.

I'll be able to look at your pull request early next week.
